### ISSUE TITLE
Configuring a multi-arch local registry.

### ISSUE THREAD

**Issue by boredazfcuk** [docker/buildx#383]
Hi,

Just wondering if someone could give me some tips, please?

I've been trying to get a local registry up and running so I can store some multi-arch images for distribution across my amd64/arm64 home lab cluster. Everything I've tried has failed.

The first hurdle I came up against was `http: server gave HTTP response to HTTPS client`. I managed to fix that, by configuring it as an insecure registry, with guidance from here: https://github.com/docker/buildx/issues/163#issuecomment-544714677

Unfortunately, that led me to this error: https://github.com/docker/buildx/issues/354, which seems like a dead end for the time being, as the issue is only 15 days old.

So I then decided that I needed to secure the registry and proceed that way... I generated some self-digned TLS certificates, but I then ran into this `x509: certificate signed by unknown authority` issue: https://github.com/docker/buildx/issues/80

I've tried adding the certificates to the local `/etc/docker/certs/<servername>:5000` folder, but no joy. I've tried adding them to the buildkit container, and defining them with a --config <file> but that just generates a ` failed to read /certs/ca.pem: open /certs/ca.pem: no such file or directory` error. I've connected to the container and verified the location, but still nothing. I've tied adding them to the /etc/ssl/certs/ca-certificates.crt file.

I've tried using the certificates that I secured my daemon with, I've tried changing the tlsverify setting in docker.json, all sorts.

Does anyone know if it's actually possible to do what I want? I.e. to push multi-arch builds into a local repo? Just really lost now so hoping someone here can tell me what I'm doing wrong.

Thanks.

**Comment by sys-out** [comment]
Hi,

i face the same problem as you, with a little difference : my cert was signed by a CA, Let's Encrypt.
It seems that the buildkit image does not contain a pem for this CA. So I added it myself to my builder container and the error disappeared (to be honest, i face another one, now !).
Maybe the same procedure could work for you too, but with your own pem file ?

First, create a local pem file near your Dockerfile, named "ca-cert-Lets_Encrypt_Authority_X3.pem" with this content : 
```
-----BEGIN CERTIFICATE-----
MIIEkjCCA3qgAwIBAgIQCgFBQgAAAVOFc2oLheynCDANBgkqhkiG9w0BAQsFADA/
MSQwIgYDVQQKExtEaWdpdGFsIFNpZ25hdHVyZSBUcnVzdCBDby4xFzAVBgNVBAMT
DkRTVCBSb290IENBIFgzMB4XDTE2MDMxNzE2NDA0NloXDTIxMDMxNzE2NDA0Nlow
SjELMAkGA1UEBhMCVVMxFjAUBgNVBAoTDUxldCdzIEVuY3J5cHQxIzAhBgNVBAMT
GkxldCdzIEVuY3J5cHQgQXV0aG9yaXR5IFgzMIIBIjANBgkqhkiG9w0BAQEFAAOC
AQ8AMIIBCgKCAQEAnNMM8FrlLke3cl03g7NoYzDq1zUmGSXhvb418XCSL7e4S0EF
q6meNQhY7LEqxGiHC6PjdeTm86dicbp5gWAf15Gan/PQeGdxyGkOlZHP/uaZ6WA8
SMx+yk13EiSdRxta67nsHjcAHJyse6cF6s5K671B5TaYucv9bTyWaN8jKkKQDIZ0
Z8h/pZq4UmEUEz9l6YKHy9v6Dlb2honzhT+Xhq+w3Brvaw2VFn3EK6BlspkENnWA
a6xK8xuQSXgvopZPKiAlKQTGdMDQMc2PMTiVFrqoM7hD8bEfwzB/onkxEz0tNvjj
/PIzark5McWvxI0NHWQWM6r6hCm21AvA2H3DkwIDAQABo4IBfTCCAXkwEgYDVR0T
AQH/BAgwBgEB/wIBADAOBgNVHQ8BAf8EBAMCAYYwfwYIKwYBBQUHAQEEczBxMDIG
CCsGAQUFBzABhiZodHRwOi8vaXNyZy50cnVzdGlkLm9jc3AuaWRlbnRydXN0LmNv
bTA7BggrBgEFBQcwAoYvaHR0cDovL2FwcHMuaWRlbnRydXN0LmNvbS9yb290cy9k
c3Ryb290Y2F4My5wN2MwHwYDVR0jBBgwFoAUxKexpHsscfrb4UuQdf/EFWCFiRAw
VAYDVR0gBE0wSzAIBgZngQwBAgEwPwYLKwYBBAGC3xMBAQEwMDAuBggrBgEFBQcC
ARYiaHR0cDovL2Nwcy5yb290LXgxLmxldHNlbmNyeXB0Lm9yZzA8BgNVHR8ENTAz
MDGgL6AthitodHRwOi8vY3JsLmlkZW50cnVzdC5jb20vRFNUUk9PVENBWDNDUkwu
Y3JsMB0GA1UdDgQWBBSoSmpjBH3duubRObemRWXv86jsoTANBgkqhkiG9w0BAQsF
AAOCAQEA3TPXEfNjWDjdGBX7CVW+dla5cEilaUcne8IkCJLxWh9KEik3JHRRHGJo
uM2VcGfl96S8TihRzZvoroed6ti6WqEBmtzw3Wodatg+VyOeph4EYpr/1wXKtx8/
wApIvJSwtmVi4MFU5aMqrSDE6ea73Mj2tcMyo5jMd6jmeWUHK8so/joWUoHOUgwu
X4Po1QYz+3dszkDqMp4fklxBwXRsW10KXzPMTZ+sOPAveyxindmjkW8lGy+QsRlG
PfZ+G6Z6h7mjem0Y+iWlkYcV4PIWL1iwBi8saCbGS5jN2p8M+X+Q7UNKEkROb3N6
KOqkqm57TH2H3eDJAkSnh6/DNFu0Qg==
-----END CERTIFICATE-----
```
Then create a "init_builder.sh" script with the following : 
```

#!/bin/sh
BUILDER_NAME=your-builder
export DOCKER_CLI_EXPERIMENTAL=enabled
docker buildx use ${BUILDER_NAME}
docker buildx inspect ${BUILDER_NAME} --bootstrap
BUILDER=$(docker ps | grep ${BUILDER_NAME} | cut -f1 -d' ')
docker cp ./ca-cert-Lets_Encrypt_Authority_X3.pem ${BUILDER}:/etc/ssl/certs/ca-cert-Lets_Encrypt_Authority_X3.pem
docker restart ${BUILDER}

```
Adapt the BUILDER_NAME value to your needs.
Execute _chmod +x init_builder.sh_
You may need to add _sudo_ before _docker_ commands, if the user on your host machine cannot execute _docker_ directly.

Finally, you can build your image with something like : 
```
./init_builder.sh
docker buildx build --platform linux/amd64,linux/arm64,linux/arm/v7 -t url_of_your_private_HTTPS_repo/your_name/your_image_tag:1.0.0 --push .
```

Hope it helps.
Regards

**Comment by boredazfcuk** [comment]
I forgot I’d opened this to be honest. I actually worked out what was happening and fixed it a few days later.

The issue was that I had a multi-node builder and the CA certificate needs to be installed on all nodes, not just the master.

I thought that the “master builder” offloaded the build process to the secondary nodes and uploaded the results they return to the registry itself. Turns out the secondary nodes upload directly to the registry so they need the certificate too.

I do have everything I did written down. I’ll make a gist after I’ve had my lunch and post it here for others finding this thread.
